dips/dip-169.md

```diff
@@ -0,0 +1,261 @@
+---
+dip: 169
+title: Multi-Agent Transactions
+authors: Emma Zhong, Tim Zakian, Sam Blackshear
+status: Draft
+type: Standard
+created: 04/05/2021
+issue: https://github.com/diem/dip/issues/169
+---
+
+
+# Summary
+
+This DIP describes a new scheme of user transaction: multi-agent transactions.
+
+# Abstract
+
+Currently in the Diem Framework, a transaction acts on behalf of a single on-chain account. However, there is no mechanism for multiple on-chain accounts to agree on a single atomic transaction. This DIP presents a new scheme of transactions--multi-agent transactions--which act on behalf of multiple on-chain accounts. Multi-agent transactions leverage Move’s [_`signer`_](https://developers.diem.com/docs/move/move-signer/) type to allow essentially any arbitrary atomic actions in one transaction involving multiple on-chain accounts.
+
+## Terminology:
+
+* Primary signer: This is the account that the transaction is sent from. This account’s sequence number is incremented, and is the account that pays gas. There is precisely one of these for every transaction.
+* Secondary signer: This is any account that participates in a multi-agent transaction that isn’t the primary sender. There can be 0-N of these. The 0 case is a normal transaction today. 
+
+# Motivation/Use Cases
+
+## Minting Directly to VASPs
+
+Today in the Diem Framework, we need two transactions in order to mint money through designated dealers to VASPs. The first transaction is a tiered-mint transaction sent by the treasury compliance account to mint money to a designated dealer’s account. The second transaction transfers the minted funds from the designed dealer’s account into the VASP’s account. This procedure requires two transactions due to the restriction that each transaction can only have one signer argument. In the multi-agent scheme, this restriction no longer exists and we can perform both steps in a single atomic transaction (see code sample below). This atomicity implies that designated dealers no longer need custody services for the money they temporarily hold in their accounts, making it more affordable and much quicker to operate as a designated dealer. This should make it much easier for more DDs to onboard in the Diem ecosystem, providing more options for VASPs. It is highly likely that in the future, DDs primarily mint directly on requests from VASPs.
+
+```rust
+fun mint<Coin>(
+   tc: signer,
+   dd: signer,
+   vasp_addr: address,
+   amount: u64,
+   tier_index: u64
+) {
+   // First, TC mints to DD.
+   tiered_mint<Coin>(
+       tc, address_of(dd), amount, tier_index
+   );
```

Comment:
Yes, I think this code sample is using existing transactions available today (since `tiered_mint` requires the received to be a DD). However, I think for the sake of this proposal let's show that we can do this with a single move function call
`mint<Coin>(tc, dd, vasp_addr, amount)` which would mint directly to vasp.
It's OK to show this conslidation of functions here imo

Reply:
+1 from me on @moeziniaf 's suggestion, it is not clear that dd's would need to hold balances on-chain in the future